communication/src/protocol.cpp

```diff
@@ -501,16 +501,66 @@ void Protocol::build_describe_message(Appender& appender, int desc_flags)
 		if (descriptor.append_system_info && (desc_flags & DESCRIBE_SYSTEM))
 		{
 			if (has_content)
+			{
 				appender.append(',');
+			}
 			has_content = true;
 			descriptor.append_system_info(append_instance, &appender, nullptr);
 		}
 		appender.append('}');
 	}
 }
 
+ProtocolError Protocol::post_description(int desc_flags)
+{
+	Message message;
+	channel.create(message);
+	const size_t header_size = Messages::describe_post_header(message.buf(), message.capacity(), 0, (desc_flags & 0xFF));
+
+	BufferAppender appender((message.buf() + header_size), (message.capacity() - header_size));
+	build_describe_message(appender, desc_flags);
+
+	const size_t msglen = (appender.next() - (uint8_t*) message.buf());
+	message.set_length(msglen);
+	if (appender.overflowed()) {
+		LOG(ERROR, "Describe message overflowed by %d bytes", appender.overflowed());
+		// There is no point in continuing to run, the device will be constantly reconnecting
+		// to the cloud. It's better to clearly indicate that the describe message is never going
+		// to go through to the cloud by going into a panic state, otherwise one would have to
+		// sift through logs to find 'Describe message overflowed by %d bytes' message to understand
+		// what's going on.
+		SPARK_ASSERT(!appender.overflowed());
+	}
+
+	LOG(INFO,"Posting '%s%s%s' describe message", desc_flags & DESCRIBE_SYSTEM ? "S" : "",
+												  desc_flags & DESCRIBE_APPLICATION ? "A" : "",
+												  desc_flags & DESCRIBE_METRICS ? "M" : "");
+
+	const ProtocolError error = channel.send(message);
+	if (error == NO_ERROR && descriptor.app_state_selector_info
```

Comment:
What is the `descriptor.app_state_selector_info`? This was copied from the other describe message and is not currently being used when posting the `DESCRIBE_METRICS` message.

I'm guessing it should probably be eliminated, but I wasn't sure if it needs to happen each time we send `(desc_flags & DESCRIBE_APPLICATION || desc_flags & DESCRIBE_SYSTEM)`.

@m-mcgowan Can you shine some light on this?

Reply:
Although it's unused, I have no problem with it being there, apart from that it is copied and pasted. I'd prefer the code to be factored out.

Reply:
The duplicate code has been refactored into a function to be called aposteriori.